File: scripts/bench_autoencoder.py
#!/usr/bin/env python3
"""
Python TMU autoencoder benchmark — mirrors bench_autoencoder.rs output format.

Modes:
  python scripts/bench_autoencoder.py            # large scale (throughput)
  python scripts/bench_autoencoder.py --small    # small scale (accuracy check)
  python scripts/bench_autoencoder.py --both     # both configs sequentially
"""
import sys
import argparse
import time

try:
    import numpy as np
except ImportError:
    sys.exit("ERROR: numpy is required.  Install with: pip install numpy")

try:
    from tmu.models.autoencoder.vanilla import TMAutoEncoder
except ImportError:
    sys.exit(
        "ERROR: tmu is not installed.\n"
        "Install with:  pip install tmu\n"
        "  or from source: pip install git+https://github.com/cair/tmu.git"
    )

# ── Configs ──────────────────────────────────────────────────────────────────
#
# clauses_per_output matches Rust's TMAutoEncoder::with_config() second argument.
# TMAutoEncoder takes number_of_clauses as the per-output count (unlike TMClassifier
# which takes the total across all classes).

LARGE = dict(
    label               = "large (throughput)",
    n_features          = 200,
    clauses_per_output  = 50,
    threshold           = 200,
    s                   = 2.0,
    state_bits          = 8,
    n_train             = 2_000,
    n_warmup            = 2,
    n_bench             = 8,
    seed                = 42,
    print_accuracy      = False,
)

SMALL = dict(
    label               = "small (accuracy check)",
    n_features          = 20,
    clauses_per_output  = 40,
    threshold           = 20,
    s                   = 3.9,
    state_bits          = 8,
    n_train             = 2_000,
    n_warmup            = 0,
    n_bench             = 20,
    seed                = 42,
    print_accuracy      = True,
)


def make_binary_data(n_train: int, n_features: int, seed: int):
    """Random i.i.d. Bernoulli(0.5) binary matrix, no labels needed."""
    rng = np.random.default_rng(seed)
    xs = rng.integers(0, 2, size=(n_train, n_features), dtype=np.uint32)
    return xs


def run_bench(cfg: dict) -> dict:
    label              = cfg["label"]
    n_features         = cfg["n_features"]
    clauses_per_output = cfg["clauses_per_output"]
    threshold          = cfg["threshold"]
    s                  = cfg["s"]
    state_bits         = cfg["state_bits"]
    n_train            = cfg["n_train"]
    n_warmup           = cfg["n_warmup"]
    n_bench            = cfg["n_bench"]
    seed               = cfg["seed"]
    print_accuracy     = cfg["print_accuracy"]

    xs = make_binary_data(n_train, n_features, seed)

    # clause_updates_per_epoch = n_train × n_features × clauses_per_output
    # (all n_features outputs are updated per sample)
    clause_updates_per_epoch = n_train * n_features * clauses_per_output

    ae = TMAutoEncoder(
        number_of_clauses       = clauses_per_output,
        T                       = threshold,
        s                       = s,
        number_of_state_bits_ta = state_bits,
        weighted_clauses        = True,
        output_active           = np.arange(n_features),
        platform                = "CPU",
    )

    print(f"\nMode   : Python TMU  [{label}]")
    print(f"Config : {n_features} features · {clauses_per_output} clauses/output · "
          f"T={threshold} · s={s} · {n_train} training samples")
    print(f"Workload: {clause_updates_per_epoch // 1_000_000} M clause updates per epoch\n")

    header = f"{'epoch':>5}  {'ms':>9}  {'samples/s':>13}  {'Mclause-ups/s':>15}"
    if print_accuracy:
        header += f"  {'recon-acc':>9}"
    print(header)

    for _ in range(n_warmup):
        ae.fit(xs, number_of_examples=n_train)

    times_s = []
    for epoch in range(n_bench):
        t0 = time.perf_counter()
        ae.fit(xs, number_of_examples=n_train)
        elapsed_s = time.perf_counter() - t0
        times_s.append(elapsed_s)

        ms   = elapsed_s * 1_000.0
        sps  = n_train / elapsed_s
        mcps = clause_updates_per_epoch / elapsed_s / 1e6

        line = f"{epoch:>5}  {ms:>8.1f}  {sps:>13.0f}  {mcps:>15.1f}"
        if print_accuracy:
            recon = ae.predict(xs)
            acc = float(np.mean(recon == xs))
            line += f"  {acc:>9.4f}"
        print(line)

    times_s.sort()
    median_s = times_s[len(times_s) // 2]
    mean_s   = sum(times_s) / len(times_s)
    min_s    = times_s[0]
    max_s    = times_s[-1]

    print()
    print(f"── Summary ({n_bench} timed epochs) {'─' * 50}")
    print(f"  median {median_s*1000:7.1f} ms  |  mean {mean_s*1000:7.1f} ms  "
          f"|  min {min_s*1000:7.1f} ms  |  max {max_s*1000:7.1f} ms")
    print(f"  throughput  : {n_train / median_s:9.0f} samples/s       "
          f"{clause_updates_per_epoch / median_s / 1e6:7.1f} Mclause-updates/s")

    return {
        "label":    label,
        "median_s": median_s,
        "mean_s":   mean_s,
        "min_s":    min_s,
        "max_s":    max_s,
        "sps":      n_train / median_s,
        "mcps":     clause_updates_per_epoch / median_s / 1e6,
    }


def main():
    parser = argparse.ArgumentParser(
        description="Python TMU autoencoder benchmark — mirrors bench_autoencoder.rs output.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    group = parser.add_mutually_exclusive_group()
    group.add_argument("--small", action="store_true",
                       help="Run small (accuracy check) config only")
    group.add_argument("--both",  action="store_true",
                       help="Run both small then large configs")
    args = parser.parse_args()

    if args.small:
        run_bench(SMALL)
    elif args.both:
        run_bench(SMALL)
        run_bench(LARGE)
    else:
        run_bench(LARGE)


if __name__ == "__main__":
    main()

File: examples/bench_autoencoder.rs
//! TMAutoEncoder throughput and accuracy benchmark.
//!
//! Runs two configs sequentially:
//!   1. Small (accuracy): 20-feature, prints reconstruction accuracy per epoch
//!   2. Large (throughput): 200-feature, measures Mclause-updates/s
//!
//! Run:
//!   cargo run --release                     --example bench_autoencoder
//!   cargo run --release --features parallel --example bench_autoencoder

use std::time::{Duration, Instant};
use tmu_rs::{Encoder, Rng, TMAutoEncoder};

struct Config {
    label: &'static str,
    n_features: usize,
    clauses_per_output: usize,
    threshold: i32,
    s: f64,
    n_train: usize,
    n_warmup: usize,
    n_bench: usize,
    print_accuracy: bool,
}

const SMALL: Config = Config {
    label: "small (accuracy check)",
    n_features: 20,
    clauses_per_output: 40,
    threshold: 20,
    s: 3.9,
    n_train: 2_000,
    n_warmup: 0,
    n_bench: 20,
    print_accuracy: true,
};

const LARGE: Config = Config {
    label: "large (throughput)",
    n_features: 200,
    clauses_per_output: 50,
    threshold: 200,
    s: 2.0,
    n_train: 2_000,
    n_warmup: 2,
    n_bench: 8,
    print_accuracy: false,
};

fn run_bench(cfg: &Config) {
    let Config {
        label,
        n_features,
        clauses_per_output,
        threshold,
        s,
        n_train,
        n_warmup,
        n_bench,
        print_accuracy,
    } = cfg;

    let mut rng = Rng::new(42);
    let xs: Vec<Vec<u8>> = (0..*n_train)
        .map(|_| (0..*n_features).map(|_| (rng.next_u64() & 1) as u8).collect())
        .collect();
    let xs_ref: Vec<&[u8]> = xs.iter().map(|v| v.as_slice()).collect();

    let encoder = Encoder::for_binary(*n_features);
    let packed = encoder.encode_batch(&xs_ref);

    let mut ae = TMAutoEncoder::with_config(
        *n_features,
        *clauses_per_output,
        *threshold,
        *s,
        8,
        true,
        42,
    );

    // Each epoch = n_train samples × n_features outputs × clauses_per_output clauses.
    let clause_updates_per_epoch = n_train * n_features * clauses_per_output;

    #[cfg(feature = "parallel")]
    let mode = "PARALLEL";
    #[cfg(not(feature = "parallel"))]
    let mode = "SEQUENTIAL";

    println!("\nMode   : {mode}  [{label}]");
    println!(
        "Config : {} features · {} clauses/output · T={} · s={} · {} training samples",
        n_features, clauses_per_output, threshold, s, n_train
    );
    println!(
        "Workload: {} M clause updates per epoch\n",
        clause_updates_per_epoch / 1_000_000
    );

    let header = if *print_accuracy {
        format!(
            "{:>5}  {:>9}  {:>13}  {:>15}  {:>8}",
            "epoch", "ms", "samples/s", "Mclause-ups/s", "recon-acc"
        )
    } else {
        format!(
            "{:>5}  {:>9}  {:>13}  {:>15}",
            "epoch", "ms", "samples/s", "Mclause-ups/s"
        )
    };
    println!("{header}");

    for _ in 0..*n_warmup {
        ae.fit_epoch(&packed);
    }

    let mut times: Vec<Duration> = Vec::with_capacity(*n_bench);
    for epoch in 0..*n_bench {
        let t = Instant::now();
        ae.fit_epoch(&packed);
        let dur = t.elapsed();
        times.push(dur);

        let ms = dur.as_secs_f64() * 1_000.0;
        let sps = *n_train as f64 / dur.as_secs_f64();
        let mcps = clause_updates_per_epoch as f64 / dur.as_secs_f64() / 1e6;

        if *print_accuracy {
            let acc = ae.reconstruction_accuracy(&packed);
            println!(
                "{:>5}  {:>8.1}  {:>13.0}  {:>15.1}  {:>7.4}",
                epoch, ms, sps, mcps, acc
            );
        } else {
            println!("{:>5}  {:>8.1}  {:>13.0}  {:>15.1}", epoch, ms, sps, mcps);
        }
    }

    times.sort();
    let median_s = times[times.len() / 2].as_secs_f64();
    let min_s = times[0].as_secs_f64();
    let max_s = times[times.len() - 1].as_secs_f64();
    let mean_s = times.iter().map(|d| d.as_secs_f64()).sum::<f64>() / times.len() as f64;

    println!();
    println!(
        "── Summary ({} timed epochs) ──────────────────────────────────────",
        n_bench
    );
    println!(
        "  median {:7.1} ms  |  mean {:7.1} ms  |  min {:7.1} ms  |  max {:7.1} ms",
        median_s * 1_000.0,
        mean_s * 1_000.0,
        min_s * 1_000.0,
        max_s * 1_000.0
    );
    println!(
        "  throughput  : {:9.0} samples/s       {:7.1} Mclause-updates/s",
        *n_train as f64 / median_s,
        clause_updates_per_epoch as f64 / median_s / 1e6
    );
}

fn main() {
    run_bench(&SMALL);
    run_bench(&LARGE);
}
